metadata-ingestion/src/datahub/integrations/airflow/operators.py

```diff
@@ -0,0 +1,55 @@
+from typing import List, Optional, Union
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+from datahub.integrations.airflow.hooks import DatahubKafkaHook, DatahubRestHook
+from datahub.metadata.com.linkedin.pegasus2avro.mxe import MetadataChangeEvent
+
+
+class DatahubBaseOperator(BaseOperator):
+    ui_color = "#4398c8"
+
+    hook: Union[DatahubRestHook, DatahubKafkaHook]
+
+    @apply_defaults
+    def __init__(
+        self,
+        *,
+        datahub_rest_conn_id: Optional[str] = None,
+        datahub_kafka_conn_id: Optional[str] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        if datahub_rest_conn_id and datahub_kafka_conn_id:
+            raise AirflowException(
+                "two hook conn id given when only exactly one required"
+            )
+        elif datahub_rest_conn_id:
+            self.hook = DatahubRestHook(datahub_rest_conn_id)
+        elif datahub_kafka_conn_id:
+            self.hook = DatahubKafkaHook(datahub_kafka_conn_id)
+        else:
+            raise AirflowException("no hook conn id provided")
+
+
+class DatahubEmitterOperator(DatahubBaseOperator):
```

Comment:
I would actually claim that this class is a bit unnecessary. Airflow is really bad at inter-task communication, and it's not recommended to send "large" chunks of data between the tasks.  Usually what you do is dump the data into for example S3 in the upstream task, and then download it and process it in the downstream task. 

The hook is valuable, as hooks are the de facto standard of interacting with external services in Airflow. So if a user want's to emit MCEs as part of their pipeline, they should be instructed use the hook inside their own custom operator.

We're going to start looking at capturing the metadata of the DAGs (DataFlow) and tasks (DataJob), either at build time or at run time. Airflow already has some [simple support](https://airflow.apache.org/docs/apache-airflow/stable/lineage.html) for capturing inputs and outputs of Tasks, which we take a look at.

Reply:
I see this PR as the first step in larger Airflow integration work. At a high level, there's three ways to get lineage and operational info:
1. Manually adding a `DatahubEmitterOperator` within your DAGs to push lineage/freshness information whenever a DAG is executed.
2. Define `inlets` and `outlets` in your DAG, let Airflow save that info in its backend via XCOM, and query it using the REST API [pull].
3. Define `inlets` and `outlets` in your DAG, and then define a LineageBackend to capture information as it happens [push]. 

I suspect we'll need (1) and (3), and might also need (2) in the interim. (1) is best for people saying "just let me easily emit lineage" and is optimized for ease of integration at the expense of architectural "cleanness". You're totally right that (3) is the best approach at a high level since it uses a "push" model - the issue is that LineageBackend is super unstable and was removed before the Airflow 2.0 release - it just got added back yesterday, after my nudge https://github.com/apache/airflow/pull/14146, but isn't yet in a release. As such, we might want to implement (2) in the interim.

As such, I'm pretty confident that doing (1) is the right first step to lay some groundwork. It's not actually doing any inter-task communication - see this [example DAG](https://github.com/linkedin/datahub/blob/5b10691e97b21b776718a416b66219766ab9a7bd/metadata-ingestion/examples/airflow/lineage_emission_dag.py).

Between (2) and (3), I'm less certain about what I should build next - I'm leaning towards (3). I believe Airflow 1.10 has support for `LineageBackend`, and it should be added back in Airflow 2.1, which is slated to be released in a few weeks.